builtin/providers/cloudstack/tags.go

```diff
@@ -0,0 +1,81 @@
+package cloudstack
+
+import (
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+// tagsSchema returns the schema to use for tags.
+//
+func tagsSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeMap,
+		Optional: true,
+	}
+}
+
+// setTags is a helper to set the tags for a resource. It expects the
+// tags field to be named "tags"
+func setTags(cs *cloudstack.CloudStackClient, d *schema.ResourceData, resourcetype string) error {
+
+	if d.HasChange("tags") {
+		oraw, nraw := d.GetChange("tags")
+		o := oraw.(map[string]interface{})
+		n := nraw.(map[string]interface{})
+		create, remove := diffTags(tagsFromMap(o), tagsFromMap(n))
```

Comment:
I would actually prefer to have ~10 lines of code than doing unneeded api calls. This way it definitely doesn't break any other api call that expect those variables to be there for the element. This tagging module will be used for any element that needs tags and they might be more fragile to removing everything and recreating it again for every change